rasa/shared/core/domain.py

```diff
@@ -252,195 +255,191 @@ def from_directory(cls, path: Text) -> "Domain":
         for form in combined.get(KEY_FORMS, []):
             if form in domain_dict.get(KEY_ACTIONS, []):
                 domain_dict[KEY_ACTIONS].remove(form)
 
-        for key in [KEY_ACTIONS, KEY_E2E_ACTIONS]:
-            duplicates[key] = self.extract_duplicates(
-                combined.get(key, []), domain_dict.get(key, [])
-            )
-            combined[key] = self.merge_lists(
-                combined.get(key, []), domain_dict.get(key, [])
-            )
-
-        for key in [KEY_FORMS, KEY_RESPONSES, KEY_SLOTS]:
-            duplicates[key] = self.extract_duplicates(
-                combined.get(key, []), domain_dict.get(key, [])
-            )
-            combined[key] = self.merge_dicts(
-                combined.get(key, {}), domain_dict.get(key, {}), override
-            )
+        combined, duplicates = Domain._merge_actions_and_e2e_actions(
+            combined, domain_dict, duplicates
+        )
+        combined, duplicates = Domain._merge_forms_responses_slots(
+            combined, domain_dict, duplicates, override
+        )
 
         if duplicates:
-            duplicates = self.clean_duplicates(duplicates)
+            duplicates = rasa.shared.utils.common.clean_duplicates(duplicates)
             combined.update({"duplicates": duplicates})
-        return combined
-
-    @staticmethod
-    def extract_duplicates(list1: List[Any], list2: List[Any]) -> List[Any]:
-        """Extracts duplicates from two lists."""
-        if list1:
-            dict1 = {
-                (sorted(list(i.keys()))[0] if isinstance(i, dict) else i): i
-                for i in list1
-            }
-        else:
-            dict1 = {}
-
-        if list2:
-            dict2 = {
-                (sorted(list(i.keys()))[0] if isinstance(i, dict) else i): i
-                for i in list2
-            }
-        else:
-            dict2 = {}
-
-        set1 = set(dict1.keys())
-        set2 = set(dict2.keys())
-        dupes = set1.intersection(set2)
-        return sorted(list(dupes))
 
-    @staticmethod
-    def clean_duplicates(dupes: Dict[Text, Any]) -> Dict[Text, Any]:
-        """Removes keys for empty values."""
-        duplicates = dupes.copy()
-        for k in dupes:
-            if not dupes[k]:
-                duplicates.pop(k)
-
-        return duplicates
-
-    @staticmethod
-    def merge_dicts(
-        tempDict1: Dict[Text, Any],
-        tempDict2: Dict[Text, Any],
-        override_existing_values: bool = False,
-    ) -> Dict[Text, Any]:
-        """Merges two dicts."""
-        if override_existing_values:
-            merged_dicts, b = tempDict1.copy(), tempDict2.copy()
+        return combined
 
-        else:
-            merged_dicts, b = tempDict2.copy(), tempDict1.copy()
-        merged_dicts.update(b)
-        return merged_dicts
+    @property
+    def data(self) -> Dict:
+        """Returns original domain dict representation."""
+        # add the config, session_config and e2e_actions defaults
```

Comment:
Can we do all of this in either the `__init__` or `from_dict` methods so it's clear in the object lifecycle when it happens?